[FSDP] Construct FQN in _full_post_state_dict_hook

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81253

Differential Revision: [D37079444](https://our.internmc.facebook.com/intern/diff/D37079444/)

fegin committed:
Construct FQN in _full_post_state_dict_hook



Differential Revision: [D37079444](https://our.internmc.facebook.com/intern/diff/D37079444/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1921,33 +1922,59 @@ def _full_post_state_dict_hook(
+        # This can happen when rank0_only is enabled and self.rank != 0.
+        if self._fsdp_wrapped_module.orig_flat_param[0] is None:
+            assert self._fsdp_wrapped_module.flat_param is not None, (
+                "When no_params is False, one of flat_param and orig_flat_param "
```
Is this a developer or user facing issue? If user can hit this issue, might be good to mention explicitly what went wrong and raise a Runtime/ValueError.

awgu replied:
It should be internal.

fegin replied:
Should be an assert here as this should not happen (anymore). If it happens, it should be an assert.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1921,33 +1922,59 @@ def _full_post_state_dict_hook(
+            module_name = module_name.replace(f"{FPW_MODULE}.", "")
+            module_name = module_name.replace(f"{FPW_MODULE}", "")
+            if module_name:
+                module_name = f"{module_name}."
```
is there a reason we need to postfix a "." here and `param_info` does not provide it like this? Seems quite easy to forget.

awgu replied:
I prefer that `param_info` keeps the `module_name` just as the module name without the `.`.

IMO, it makes more logical sense that `module_name` is just the module's name. The `.` is only relevant when prefixing for a parameter name, which may not be the only reason we access `module_name`.

fegin replied:
Will keep this statement. `param_info` should keeps the `module_name` as the actual module name.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1921,33 +1922,59 @@ def _full_post_state_dict_hook(
+
             # Clone non-ignored parameters before exiting the
             # `_summon_full_params()` context
+            key = fqn
```
just use `clean_key` here?

fegin replied:
Using `clean_key` is not correct as `clean_key` may be stripped out. Remove this key and directly use `fqn`.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1962,7 +1989,21 @@ def _full_post_state_dict_hook(
+                    f"{checkpoint_wrapper._CHECKPOINT_PREFIX}.", ""
+                )
+                fqn = f"{prefix}{clean_key}"
+                fqn = fqn.replace("mod.", "")
```
this should not be needed now?

rohan-varma reviewed (approved):
Some minor comments and "mod" prefixing should not be needed now. Thanks!

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1921,33 +1922,59 @@ def _full_post_state_dict_hook(
+        if self._fsdp_wrapped_module.orig_flat_param[0] is None:
+            assert self._fsdp_wrapped_module.flat_param is not None, (
+                "When no_params is False, one of flat_param and orig_flat_param "
+                "should has value."
```
nit:
```suggestion
                "should have value."
```

fegin committed:
Update on "[FSDP] Construct FQN in _full_post_state_dict_hook"


Differential Revision: [D37079444](https://our.internmc.facebook.com/intern/diff/D37079444/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Construct FQN in _full_post_state_dict_hook"


Differential Revision: [D37079444](https://our.internmc.facebook.com/intern/diff/D37079444/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin committed:
Update on "[FSDP] Construct FQN in _full_post_state_dict_hook"


Differential Revision: [D37079444](https://our.internmc.facebook.com/intern/diff/D37079444/)

[ghstack-poisoned]

fegin commented:
#pytorchbot merge

fegin commented:
@pytorchbot merge

fegin committed:
Update on "[FSDP] Construct FQN in _full_post_state_dict_hook"


Differential Revision: [D37079444](https://our.internmc.facebook.com/intern/diff/D37079444/)

[ghstack-poisoned]
